README.md
# Responsify  [![Build Status](https://travis-ci.org/eclifford/responsify.svg?branch=master)](https://travis-ci.org/eclifford/responsify) [![Coverage Status](https://img.shields.io/coveralls/eclifford/responsify.svg)](https://coveralls.io/r/eclifford/responsify?branch=master)

> Responsify.js is a src-N based responsive image solution optimized for
dynamic image solutions such as Adobe Scene7.

## Features

- Support for parameterized media image services such as Adobe Scene7
- Support for declarative images per breakpoint if you want to serve your own responsive images
- Built in customizable breakpoint detection/notification for use in responsive grids such as
those found in Bootstrap and Foundation
- No third party libraries required
- Non blocking script

## Why Responsify?

Responsify was born out of the need for a **Scene7** based responsive solution for a large enterprise content managed web
application. We needed something that was **fast**, **non blocking**, **standalone** and relied on **convention over declaration**.

In the end we created a solution that should work not only for **Scene7** but for any **src-N** based
responsive image solution.

## Quick Start

### Installation with Bower

```bash
bower install responsify
```

### Add Responsify to DOCUMENT

Responsify can be loaded anywhere in the `<head>` or `<body>` as either a blocking script or with **defer** and **async**. Responsify will poll the DOM prior to the loaded event to continually process images if using **defer** or if put at in the `<head>`.

```html
<head>
  <script src='bower_components/resonsify/responsify.js' async></script>
</head>
```

### Initialize Responsify

Responsify needs to be told when to initialize. 


Vanilla Face

```html
<script>
  document.addEventListener("DOMContentLoaded", function() {
    Responsify.init();
  });
</script>
```

jQuery

```html
<script>
  $(document).ready(function() {
    Responsify.init();
  });
</script>
```

### Basic Scene7 Responsive image

Responsify looks for all images by selector `img.responsive` in container element
`document` by default.

At the very least assuming you were using Scene7 you would need to populate the `data-responsify`
attribute to your Scene7 resource location.

```html
<img class='responsive'
  data-responsify="http://s7d9.scene7.com/is/image/DEMOAKQA/1440.1?wid={width}"
/>
```

### Basic Declarative Responsive Image

```html
<img class='responsive'
  data-responsify="http://www.test.com/images/"
  data-responsify-break-a="foo_320.jpg"
  data-responsify-break-b="foo_640.jpg"
  data-responsify-break-c="foo_1024.jpg"
/>
```

Assuming your on **break-b** the resulting markup.

```html
<img class='responsive'
  src="http://www.test.com/images/foo_640.jpg"
  data-responsify="http://www.test.com/images/"
  data-responsify-break-a="foo_320.jpg"
  data-responsify-break-b="foo_640.jpg"
  data-responsify-break-c="foo_1024.jpg"
/>
```

### Options

#### namespace

Type: `String` Default: **responsify**

The namespace to use for `data-attribute` prefix.

#### selector

Type: `String` Default: **img.responsive,div.responsive**

The query selector to use when searching for images or background divs to process.

#### lazyload

Type: `Boolean` Default: **true**

Whether or not to only load images that are currently visible in the viewport.

#### breakpoints

Type: `Array[Object]` Default: **Array of breakpoint objects**

Your breakpoint grid

### Events

Responsify publishes out the following events. To subscribe to these you
simply use `Responsify.on`.

```js
  Responsify.on('responsify:breakpoint:change', function() {
    // do something
  });
```

#### responsify:image:rendered

```js
  Responsify.on('responsify:image:rendered', function(el) {
    // do something
  });
```

### Contributing

Fork the repo and issue a pull request

### License

The MIT License

Copyright (c) 2014 Eric Clifford

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
<a name="0.2.2"></a>
### 0.2.2 (2015-02-06)


#### Bug Fixes

* **Logr:** better handling of downscaling of imgs and background divs ((cb37cbd3))


<a name="0.2.1"></a>
### 0.2.1 (2015-02-03)


#### Features

* **Responsify:** adding viewport lazy loading, rendering before pageload ((671a2223))


<a name="0.2.0"></a>
## 0.2.0 (2014-10-30)


<a name="0.1.1"></a>
### 0.1.1 (2014-10-28)


<a name="0.1.0"></a>
## 0.1.0 (2014-10-13)


#### Features

* **Responsify:** adding better exception handling and visible image detection ((032636e2))



Gruntfile.js
module.exports = function(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('bower.json'),
    uglify: {
      options: {
        banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' +
          '<%= pkg.author %> - <%= grunt.template.today("yyyy-mm-dd") %> */'
      },
      dist: {
        files: {
          'responsify.min.js': ['responsify.js']
        }
      }
    },
    jshint: {
      all: ['responsify.js']
    },
    watch: {
      files: ['responsify.js', 'test/**/*.js'],
      tasks: ['jshint', 'karma:unit:run']
    },
    karma: {
      options: {
        frameworks: ['mocha', 'chai', 'jquery-2.1.0', 'sinon-chai'],

        files: [
          'bower_components/MutationObserver-shim/MutationObserver.js',
          'responsify.js',
          'test/unit/**/*.js',
          'test/fixtures/**/*.html'
        ],

        preprocessors: {
          'responsify.js': ['coverage']
        }

      },
      unit: {
        reporters: ['progress', 'coverage'],
        browsers: ['PhantomJS'],
        coverageReporter: {
          type : 'html',
          dir : 'test/coverage/'
        }
      },
      ci: {
        reporters: ['progress', 'coverage', 'coveralls'],
        browsers: ['PhantomJS'],
        coverageReporter: {
          type : 'lcov',
          dir : 'test/coverage/'
        },
        singleRun: true
      }
    },
    replace: {
      dist: {
        src: ['responsify.js'],
        overwrite: true,
        replacements: [{
          from: /version:.'[0-9]+.[0-9]+.[0-9]+'/g,
          to: "version: '<%= pkg.version %>'"
        }]
      }
    },
    bump: {
      options: {
        files: ['package.json', 'bower.json'],
        updateConfigs: ['pkg'],
        pushTo: 'origin',
        commitFiles: ['-a']
      }
    },
    changelog: {
      default: {
        options: {}
      }
    }
  });

  grunt.loadNpmTasks('grunt-karma');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-text-replace');
  grunt.loadNpmTasks('grunt-bump');
  grunt.loadNpmTasks('grunt-conventional-changelog');

  grunt.registerTask('release', 'Build and release plugin', function(type) {
    grunt.task.run([
      "jshint",
      "bump-only:" + (type || "patch"),
      "replace",
      "uglify",
      "changelog",
      "bump-commit"
    ]);
  });
};


responsive.js
/*
  Responsify
  - don't block me bro ('don't require this script to be a blocking script in the head')
  - don't require third party libs (jQuery, ImagesLoaded, Underscore)
  - detect resize (debounced) and use crop parameters to select appropriate images for breakpoint
  - responsive images are the size of their containing elements
  - images added to dom post page load should be detected and rendered properly
  - images not currently visible in viewport should not be rendered
  - detect scroll (debounced) detect newly visible images and rendered them properly
  - image elements will use data-src-n approach for query parameter and will use src attribute
  for default low res versions of images
  - Will require polyfill (https://github.com/Polymer/MutationObservers) for IE9/10
*/
(function(root, factory) {

  if (typeof define === 'function' && define.amd) {
    define([], function() {
      return (root.Responsify = factory(root));
    });
  } else if (typeof exports !== 'undefined') {
    module.exports = factory(root);
  } else {
    root.Responsify = factory(root);
  }

}(this, function(root) {
  'use strict';

  var Responsify = {
    version: '0.0.1',

    debug: true,

    imagesOffScreen: [],

    images: [{
      node: {},
      visible: true
    }],

    breakpoints: ['480', '768', '1024', '1280', '1440'],

    activeBreakpoint: null,

    init: function() {
      var self = this;
      this.processImages();
      this.onImageMutationEvent(function(node) {
        self.processImage(node);
      });

      this.activeBreakpoint = this.getClosestBreakpoint();

      this.onScrollEvent(function() {
        self.processImages();
      });

      this.onResizeEvent(function() {
        var currentBreakpoint = self.getClosestBreakpoint();
        if (currentBreakpoint != self.activeBreakpoint) {
          self.activeBreakpoint = currentBreakpoint;
          self.processImages();
        }
      });
    },

    getClosestBreakpoint: function() {
      var viewport = window.innerWidth;

      for(var i = 0; i < this.breakpoints.length; i ++) {
        if(viewport <= this.breakpoints[0]) {
          return 0;
        }
        if(viewport <= this.breakpoints[i] && viewport > this.breakpoints[i-1]) {
          return i;
        }
      }
    },

    processImages: function() {
      var images = document.getElementsByClassName('responsive');

      for(var i = 0; i < images.length; i++) {
        this.processImage(images[i]);
      }
    },

    processImage: function(img) {
      if(this.isImageOnScreen(img)) {
        this.log('image processed');
        // get data-src-1
        var src = img.dataset.src;
        var params = img.dataset["src-" + (this.activeBreakpoint + 1)];
        var width = img.parentElement.clientWidth;
        img.src = src + "&wid=" + width;
      }
      // img.src = img.dataset.src += '?resMode=sharp2&wid=' + img.parentElement.clientWidth;
    },

    processVisibleImages: function() {
      var visibleImages = this.images.filter(function(img) {
        return img.visible;
      });

      visibleImages.forEach(function(img) {
        img.element.src = img.element.src += '?resMode=sharp2&wid=' + img.element.parentElement.clientWidth;
      });
    },

    isImageOnScreen: function(element) {
      var elementRect = element.getBoundingClientRect();
      var viewportHeight = document.body.clientHeight;

      if ((elementRect.top >= 0 && elementRect.top <= viewportHeight) || (elementRect.bottom >= 0 && elementRect.bottom <= viewportHeight))
        return true;
      else
        return false;
    },

    onScrollEvent: function(callback) {
      var self = this;
      window.addEventListener("scroll", this.debounce(function() {
        callback();
      }, 500));
    },

    onImageMutationEvent: function(callback) {
      var MutationObserver = window.MutationObserver || window.WebKitMutationObserver || window.MozMutationObserver;

      var observer = new MutationObserver(function (mutations) {
        mutations.forEach(function (mutation) {
          console.log('mutation');
          if (mutation.addedNodes) {
            for (var i = 0; i < mutation.addedNodes.length; i++) {
              var node = mutation.addedNodes[i];
              if (node.nodeType == 1 && node.tagName == "IMG") {
                callback(node);
              }
            }
          }
        });
      });

      observer.observe(document, {
        childList: true,
        subtree: true
      });
    },

    // Listen for resize events
    onResizeEvent: function(callback) {
      var self = this;
      window.addEventListener("resize", this.debounce(function() {
        self.log('resize event fired');
        callback();
      }, 500));
    },


    log: function(message) {
      if(this.debug) {
        console.log(message);
      }
    },

    debounce: function(func, wait, immediate) {
      var timeout, args, context, timestamp, result;

      var later = function() {
        var last = Date.now - timestamp;
        if (last < wait) {
          timeout = setTimeout(later, wait - last);
        } else {
          timeout = null;
          if (!immediate) {
            result = func.apply(context, args);
            context = args = null;
          }
        }
      };

      return function() {
        context = this;
        args = arguments;
        timestamp = Date.now;
        var callNow = immediate && !timeout;
        if (!timeout) {
          timeout = setTimeout(later, wait);
        }
        if (callNow) {
          result = func.apply(context, args);
          context = args = null;
        }

        return result;
      };
    }
  };

  return Responsify;
}));


responsify.js
/*!
 * Responsify
 * http://github.com/eclifford/responsify
 *
 * Author: Eric Clifford
 * Email: ericgclifford@gmail.com
 * Date: 10.10.2014
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    define([], function() {
      return (root.Responsify = factory(root));
    });
  } else if (typeof exports !== 'undefined') {
    module.exports = factory(root);
  } else {
    root.Responsify = factory(root);
  }
}(this, function(root) {
  'use strict';

  var Responsify = {
    version: '0.2.2',

    // default options
    options: {
      namespace: 'responsify',
      selector: 'img.responsive,div.responsive',
      lazyload: false,
      breakpoints: [
        {
          label: 'break-a',
          enter: 0,
          exit: 765
        },
        {
          label: 'break-b',
          enter: 768,
          exit: 991
        },
        {
          label: 'break-c',
          enter: 992,
          exit: 1199
        },
        {
          label: 'break-d',
          enter: 1200,
          exit: 10000
        }
      ]
    },

    currentBreakpoint: null,

    images: [],

    events: {},

    interpolations: {},

    /**
     * Initialize Responsify
     *
     * @example
     *     Responsify.init()
     *
     * @param {Object} config the configuration object to override defaults
     * @api public
     */
    init: function(config) {
      var self = this;

      // extend responsify options with passed in configuration
      if (config) this.extend(this.options, config);

      // setup our default interpolations
      this.setupDefaultInterpolations();

      // get the current breakpoint
      this.currentBreakpoint = this.findClosestBreakpoint(window.innerWidth);

      // register all events
      this.setupEvents();

      // poll dom for new images until page load
      var intervalId = setInterval( function onPreInit() {
        self.addImages(self.getNewImages());

        if ( /^loaded|^i|^c/.test( window.document.readyState ) ) {
          clearInterval( intervalId );

          // for any added/removed mutation we reset/refresh
          var observer = new MutationObserver(function(mutations) {
            window.requestAnimationFrame(function onMutationEvent() {
              self.addImages(self.getNewImages());
            });
          });

          observer.observe(window.document.body, {
            childList: true,
            subtree: true
          });

          return;
        }
      }, 100);
    },
    /**
     * Setup our default string interpolation methods
     *
     * @example
     *     Responsify.setupDefaultInterpolations()
     *
     * @api public
     */
    setupDefaultInterpolations: function() {
      this.interpolations.width = function(el) {
        return el.parentElement.clientWidth;
      };
      this.interpolations.landscape = function(el) {
        return Math.ceil(el.parentElement.clientWidth * 0.5625);
      };
      this.interpolations.height = function(el) {
        return el.parentElement.clientHeight;
      };
    },
    /**
     * Setup event handlers for Responsify lifecycle
     *
     * @example
     *     Responsify.setupEvents()
     *
     * @api public
     */
    setupEvents: function() {
      var self = this;
      window.addEventListener("resize", function() {
        window.requestAnimationFrame(function onRequestResizeEvent() {
          self.onResizeEvent(window.innerWidth);
        });
      });

      if (this.options.lazyload) {
        window.addEventListener("scroll", function() {
          window.requestAnimationFrame(function() {
            self.refreshImages();
          });
        });
      }
    },
    /**
     * Get all images added to DOM since last check and return them
     *
     * @example
     *     Responsify.getNewImages()
     *
     * @api public
     */
    getNewImages: function() {
      var self = this,
          imgs = [].slice.call(window.document.querySelectorAll(this.options.selector));

      return imgs.filter(function(el) {
        return self.images.indexOf(el) === -1;
      });
    },
    /**
     * Rerender all images currently stored
     *
     * @example
     *     Responsify.refreshImages()
     *
     * @api public
     */
    refreshImages: function() {
      this.renderImages(this.images);
    },
    /**
     * Find all images on the DOM that match selector and store them
     *
     * @example
     *     Responsify.resetImages()
     *
     * @api public
     */
    resetImages: function() {
      this.images = [].slice.call(window.document.querySelectorAll(this.options.selector));
    },
    /**
     * Upon resize test for breakpoint change and re-render images
     *
     * @example
     *     Responsify.onResizeEvent(100)
     *
     * @param {Number} width the width of the current window
     * @api public
     */
    onResizeEvent: function(width) {
      var breakpoint = null;

      if (isNaN(width))
        throw Error("Responsify.onResizeEvent(): expects parameter width of type Number");

      breakpoint = this.findClosestBreakpoint(width);
      if (breakpoint != this.currentBreakpoint) {
        this.setBreakpoint(breakpoint);

        this.refreshImages();
      }
    },
    /**
     * Set the current active breakpoint to the one supplied
     *
     * @example
     *     Responsify.setBreakpoint({Object})
     *
     * @param {Object} breakpoint the breakpoint to set
     * @api public
     */
    setBreakpoint: function(breakpoint) {
      if (typeof breakpoint !== 'object')
        throw new Error("Responsify.setBreakpoint(): expects parameter breakpoint of type Object");

      this.currentBreakpoint = breakpoint;
      this.publish('responsify:breakpoint:change', this.currentBreakpoint);
    },
    /**
     * Return the closest breakpoint for the supplied width
     *
     * @example
     *     Responsify.findClosestBreakpoint(100)
     *
     * @param {Number} width the width to test for a breakpoint
     * @return {Object} the breakpoint that matches supplied width
     * @api public
     */
    findClosestBreakpoint: function(width) {
      if (isNaN(width))
        throw new Error("Responsify.findClosestBreakpoint(): expects parameter width of type Number");

      for (var i = 0; i < this.options.breakpoints.length; i++) {
        if (width >= this.options.breakpoints[i].enter && width <= this.options.breakpoints[i].exit) {
          return this.options.breakpoints[i];
        }
      }
    },
    /**
     * Return whether or not the supplied breakpoint is current
     *
     * @example
     *     Responsify.isBreakpointEqualTo('break-a')
     *
     * @param {String} breakpoint the breakpoint label to test
     * @return {Boolean} whether or not the breakpoint is active
     * @api public
     */
    isBreakpointEqualTo: function(breakpoint) {
      if (typeof breakpoint !== 'string')
        throw new Error("Responsify.isBreakpointEqualTo(): expects parameter breakpoint of type String");

      return breakpoint === this.currentBreakpoint.label;
    },
    /**
     * Return whether or not the supplied HTMLElement is visible or not
     *
     * @example
     *     Responsify.isElementVisible(<HTMLElement>)
     *
     * @param {HTMLElement} the element to test
     * @return {Boolean} whether or not the element is visible
     * @api public
    */
    isElementVisible: function(el) {
      if (!(el && el instanceof HTMLElement))
        throw new Error("Responsify.isElementVisible(): expects parameter el of type HTMLElement");

      return !!el.offsetParent;
    },
    /**
     * Return whether or not element is in viewport
     *
     * @example
     *     Responsify.isElementInViewport(<HTMLElement>)
     *
     * @param {HTMLElement} the element to test
     * @return {Boolean} whether or not the element is in viewport
     * @api public
    */
    isElementInViewport: function(el) {
      if (!(el && el instanceof HTMLElement))
        throw new Error("Responsify.isElementInViewport(): expects parameter el of type HTMLElement");

      var rect = el.getBoundingClientRect();
      var html = document.documentElement;
      return (
        rect.top <= (window.innerHeight || html.clientHeight) &&
        rect.left >= 0 &&
        rect.right <= (window.innerWidth || html.clientWidth)
      );
    },
    /**
     * Return whether or not image src would be downscaled
     *
     * @example
     *     Responsify.isElementToBeDownscaled(<HTMLElement>)
     *
     * @param {HTMLElement} the element to test
     * @return {Boolean} whether or not the image could be downscaled
     * @api public
    */
    isElementToBeDownscaled: function(el) {
      var url = "",
          img;

      // typecheck
      if (!(el && el instanceof HTMLElement))
        throw new Error("Responsify.isElementBeingDownscaled(): expects parameter el of type HTMLElement");

      // element is an image
      if (el.nodeName.toLowerCase() === 'img') {
        if (el.naturalWidth > el.parentElement.clientWidth)
          return true;
      }
      // element is a div
      else if (el.nodeName.toLowerCase() === 'div'){
        // extract real url from style syntax
        url = /^url\((['"]?)(.*)\1\)$/.exec(el.style.backgroundImage);
        url = url ? url[2] : "";
        img = new Image();
        img.src = url;

        if (img.naturalWidth > el.parentElement.clientWidth)
          return true;

      } else {
        throw new Error("Responsify.isElementBeingDownscaled(): element is not a div or an img");
      }
    },
    /**
     * Render an Array or NodeList of images that are visible
     *
     * @example
     *     Responsify.renderImages([<HTMLElement>])
     *
     * @param {Array|NodeList} images the images to render
     * @api public
     */
    renderImages: function(images) {
      if (!images)
        throw new Error("Responsify.renderImages(): expects parameter images of type Array or NodeList");

      for (var i = 0; i < images.length; i++) {
        this.renderImage(images[i]);
      }
    },
    /**
     * Render an image or background div
     *
     * @example
     *     Responsify.renderImage(<HTMLElement>)
     *
     * @param {HTMLElement} el the element to render
     * @return {Boolean} whether or not the image was rendered
     * @api public
     */
    renderImage: function(el) {
      var baseURI = "",
          breakpointURI = "",
          imageURI = "",
          baseWidth = 0,
          baseHeight = 0,
          ratio = 1,
          width = 0,
          height = 0;

      if (!(el && el instanceof HTMLElement))
        throw new Error("Responsify.renderImage(): expects parameter el of type HTMLElement");

      // visibility test
      if (!this.isElementVisible(el))
        return false;

      // viewport test
      if (this.options.lazyload && !this.isElementInViewport(el))
        return false;

      // downscale test
      if (this.isElementToBeDownscaled(el))
        return false;

      // read in base URI and current breakpoint URI from data attributes
      baseURI = el.getAttribute('data-' + this.options.namespace) || "";
      breakpointURI = el.getAttribute("data-" + this.options.namespace + "-" + this.currentBreakpoint.label) || "";

      // combine baseURI and breakpoint data
      imageURI = this.buildImageURI(baseURI, breakpointURI);

      // interpolate string running it through all interpolation methods to create dynamic values
      imageURI = this.getURLInterpolationsForElement(imageURI, el);

      // render image or background image div
      if (el.nodeName.toLowerCase() === 'img') {
        if (el.src !== imageURI)
          el.src = imageURI;
      } else {
        el.style.backgroundImage = "url('" + imageURI + "')";
        el.style.backgroundSize = "cover";
      }

      // notify subscribers
      this.publish('responsify:image:rendered', el);

      return true;
    },
    /**
     * Interpolate all placeholders by applying interpolation function
     * on el
     *
     * @example
     *     Responsify.getURLInterpolationsForElement(
     *       'http://http://s7d9.scene7.com/is/image/DEMOAKQA/1440.1?wid={width}',
     *       <HTMLElement>
     *     )
     *
     * @param {String} url the url to interpolate
     * @param {HTMLElement} el to pass to interpolation method
     * @api public
     */
    getURLInterpolationsForElement: function(url, el) {
      var re = /{([^%}]+)?}/g,
          match = null;

      // for each match replace placeholder with interpolated value
      while ((match = re.exec(url)) !== null) {
        url = url.replace(match[0], this.interpolations[match[1]](el));
      }

      return url;
    },
    /**
     * Combine baseURI with current breakpoint parameters
     *
     * @example
     *     Responsify.buildImageURI("http://www.test.com", "?foo=test")
     *
     * @param {String} baseURI the base uri
     * @param {String} breakpointURI the breakpoint uri to append
     * @return {String} the built uri
     * @api public
     */
    buildImageURI: function(baseURI, breakpointURI) {
      var uri = "",
          obj = {};

      if (typeof baseURI !== 'string')
        throw new Error("Responsify.buildImageURI(): expects parameter baseURI of type String");

      if (typeof breakpointURI !== 'string')
        throw new Error("Responsify.buildImageURI(): expects parameter breakpointURI of type String");

      // if breakpoint is full url just return it
      if (breakpointURI.match(/(http|https)/g)) {
        return breakpointURI;
      }
      // querystring or path + querystring
      else if (breakpointURI.indexOf('?') !== -1) {
        obj = breakpointURI.split('?');
        uri = this.appendQueryString(baseURI + obj[0], obj[1]);
      }
      // just querystring
      else if (breakpointURI.indexOf('=') !== -1) {
        uri = this.appendQueryString(baseURI, breakpointURI);
      }
      // just path
      else {
        uri = baseURI + breakpointURI;
      }
      return uri;
    },
    /**
     * Append query string to passed base URI
     *
     * @example
     *     Responsify.appendQueryString("http://www.test.com", "?foo=test")
     *
     * @param {String} uri the base uri
     * @param {String} query the querystring to add
     * @return {String} the built uri
     * @api public
     */
    appendQueryString: function(uri, query) {
      var seperator;

      if (typeof uri !== 'string')
        throw new Error("Responsify.appendQueryString(): expects parameter uri of type String");

      if (!query)
        return uri;

      query = query.replace(/^(&|\?)/, '');
      seperator = uri.indexOf('?') !== -1 ? '&' : '?';
      return uri + seperator + query;
    },
    /**
     * Add an image to internal managed array
     *
     * @example
     *     Responsify.addImage(<img>)
     *
     * @param {HTMLElement} img the image to add
     * @api public
     */
    addImage: function(img) {
      if (!(img && img instanceof HTMLElement))
        throw new Error("Responsify.addImage(): expects parameter img of type HTMLElement");

      this.images.push(img);
      this.renderImage(img);
    },
    /**
     * Add an array of images to internal managed array
     *
     * @example
     *     Responsify.addImages([<img>, <img>])
     *
     * @param {Array} imgs the images to add
     * @api public
     */
    addImages: function(imgs) {
      for(var i = 0; i < imgs.length; i++) {
        this.addImage(imgs[i]);
      }
    },
    /**
     * Remove supplied image from internal managed array
     *
     * @example
     *     Responsify.removeImage(<img>)
     *
     * @param {HTMLElement} img the image to remove
     * @api public
     */
    removeImage: function(img) {
      if (!(img && img instanceof HTMLElement))
        throw new Error("Responsify.removeImages(): expects parameter img of type HTMLElement");
      var i = this.images.length;
      while (i--) {
        if (img === this.images[i]) {
          this.images.splice(i, 1);
        }
      }
    },
    /**
     * Remove images from internal managed array
     *
     * @example
     *     Responsify.removeImages([<img>,<img>])
     *
     * @param {Array} imgs the images to remove
     * @api public
     */
    removeImages: function(imgs) {
      if (Object.prototype.toString.call(imgs) !== "[object NodeList]")
        throw new Error("Responsify.removeImages(): expects parameter imgs of type NodeList");

      for (var i = 0; i < imgs.length; i++) {
        this.removeImage(imgs[i]);
      }
    },
    /**
     * Simple publish method for notifing subscribers of an event
     *
     * @example
     *     Responsify.publish('event:name');
     *
     * @param {String} topic the topic to publish
     * @api public
     */
    publish: function(topic) {
      var subs, len;

      if (typeof topic !== 'string')
        throw new Error("Responsify.publish(): expects parameter topic of type String");

      subs = this.events[topic];
      len = subs ? subs.length : 0;

      //can change loop or reverse array if the order matters
      while (len--) {
        subs[len].handler.apply(subs[len].context, [].slice.call(arguments, 1));
      }
    },
    /**
     * Simple subscribe method for adding handlers to event topics.
     *
     * @example
     *     var handler = function() {};
     *     Responsify.on('event:name', handler);
     *
     * @param {String} topic the topic to remove handler from
     * @param {Function} handler the function to remove
     * @param {Object} context the context of `this` in which the handler should execute
     * @api public
     */
    on: function(topic, handler, context) {
      if (typeof topic !== 'string')
        throw new Error("Responsify.on(): expects parameter topic of type String");

      if (typeof handler !== 'function')
        throw new Error("Responsify.on(): expects parameter handler to of type Function");

      if (!this.events[topic]) {
        this.events[topic] = [];
      }
      this.events[topic].push({
        handler: handler,
        context: context || this
      });
    },
    /**
     * Simple unsubscribe method for removing subscribers by stored handler from
     * internal events hash.
     *
     * @example
     *     var handler = function() {};
     *     Responsify.off('event:name', handler);
     *
     * @param {String} topic the topic to remove handler from
     * @param {Function} handler the function to remove
     * @api public
     */
    off: function(topic, handler) {
      var i = 0;

      if (typeof topic !== 'string')
        throw new Error("Responsify.off(): expects parameter topic of type String");

      if (typeof handler !== 'function')
        throw new TypeError("Responsify.off(): expects parameter handler of type Function");

      if (this.events[topic]) {
        i = this.events[topic].length;

        while (i--) {
          if (!handler || this.events[topic][i].handler === handler) {
            this.events[topic].splice(i, 1);
          }
        }
      }
    },
    /**
     * Extend the destination object by coping over all properties recursively from
     * source object.
     *
     * @example
     *     Responsify.extend({ foo: 'bar'}, { baz: 'woz'});
     *     // => { foo: 'bar', baz: 'woz' }
     *
     * @param {Object} destination object to extend to
     * @param {Object} source object to extend from
     * @return {Object} the extended destination object
     * @api public
     */
    extend: function(destination, source) {
      if (typeof destination !== 'object')
        throw new Error("Responsify.extend(): expects parameter destination of type Object");

      if (typeof source !== 'object')
        throw new Error("Responsify.extend(): expects parameter source of type Object");

      // copy over all properties from source object to destination object
      for (var prop in source) {
        // if property is an object but not an array recursively copy
        if (typeof source[prop] === 'object' && Object.prototype.toString.call(source[prop]) !== '[object Array]') {
          destination[prop] = this.extend(destination[prop], source[prop]);
        } else {
          destination[prop] = source[prop];
        }
      }
      return destination;
    }
  };

  // polyfills
  window.requestAnimationFrame = window.requestAnimationFrame || window.mozRequestAnimationFrame || window.webkitRequestAnimationFrame || function (callback) { window.setTimeout(callback, 1000 / 60); };
  window.MutationObserver = window.MutationObserver || window.WebKitMutationObserver || window.MozMutationObserver;

  return Responsify;
}));
